bigflow/build/dist.py

```diff
@@ -27,74 +21,31 @@
 import bigflow.version
 import bigflow.build.pip
 import bigflow.build.dev
-import bigflow.build.dataflow.dependency_checker
-import bigflow.commons as bf_commons
+import bigflow.build.operate
+import bigflow.build.spec as spec
 
 
 logger = logging.getLogger(__name__)
 
 SETUP_VALIDATION_MESSAGE = 'BigFlow setup is valid.'
 
 
-def run_tests(project_dir: Path, build_dir: Path, test_package: Path):
-    output_dir = build_dir / 'junit-reports'
-    try:
-        return bf_commons.run_process([
-            "python", "-m", "xmlrunner", "discover",
-            "-s", test_package,
-            "-t", project_dir,
-            "-o", output_dir,
-        ])
-    except subprocess.CalledProcessError:
-        raise ValueError("Test suite failed.")
-
-
-def export_docker_image_to_file(tag: str, target_dir: Path, version: str):
-    image_target_path = target_dir / f"image-{version}.tar"
-    logger.info("Exporting the image to %s ...", image_target_path)
-    bf_commons.run_process(["docker", "image", "save", "-o", image_target_path, bf_commons.get_docker_image_id(tag)])
-
-
-def build_docker_image(project_dir: Path, tag: str):
-    print('Building a Docker image. It might take a while.')
-    bf_commons.run_process(f'docker build {project_dir} --tag {tag}')
-
-
-def build_dags(
-        root_package: Path,
-        project_dir: Path,
-        docker_repository: str,
-        start_time: str,
-        version: str,
-        specific_workflow: typing.Optional[str] = None):
-    for workflow in bigflow.cli.walk_workflows(root_package):
-        if specific_workflow is not None and specific_workflow != workflow.workflow_id:
-            continue
-        print(f'Generating DAG file for {workflow.workflow_id}')
-        bigflow.dagbuilder.generate_dag_file(
-            str(project_dir),
-            docker_repository,
-            workflow,
-            start_time,
-            version,
-            root_package.name)
-
-
-def build_image(
-        docker_repository: str,
-        version: str,
-        project_dir: Path,
-        image_dir: Path,
-        deployment_config: Path):
-
-    os.mkdir(image_dir)
-    tag = bf_commons.build_docker_image_tag(docker_repository, version)
-    build_docker_image(project_dir, tag)
-    try:
-        export_docker_image_to_file(tag, image_dir, version)
-        shutil.copyfile(deployment_config, image_dir / deployment_config.name)
-    finally:
-        bf_commons.remove_docker_image_from_local_registry(tag)
+class BigflowDistribution(distutils.dist.Distribution):
+    """Customized Distribution for bigflow projects. Add custom commands, allow access to bigflow project spec."""
+
+    def __init__(self, attrs=None):
```

Comment:
Type hints please, I'm wondering why there is `attrs = dict(attrs or {})` below

Reply:
Because `attrs` might be None according to 'dist.Distribution' sources.